File: ocr_reader.rb
# ocr_reader.rb — Ruby версия

require 'rtesseract'
require 'optparse'
require 'time'

options = {}
OptionParser.new do |opts|
  opts.banner = "Usage: ruby ocr_reader.rb <image> [-l lang] [-o output]"
  opts.on("-l", "--lang LANG", "Язык OCR (по умолчанию eng)") { |l| options[:lang] = l }
  opts.on("-o", "--output FILE", "Файл для сохранения результата") { |o| options[:output] = o }
end.parse!

image_path = ARGV[0]
unless image_path && File.exist?(image_path)
  puts "❌ Файл не найден: #{image_path}"
  exit 1
end

lang = options[:lang] || 'eng'

puts "\e[36m📖 OCR Reader (Ruby)\e[0m"
puts "📂 Обработка: #{image_path}"
puts "⏳ Распознавание..."

start = Time.now
begin
  text = RTesseract.new(image_path, lang: lang).to_s.strip
  elapsed = Time.now - start

  puts "\n\e[32mРезультат:\e[0m"
  puts "─" * 50
  puts text
  puts "─" * 50

  words = text.split(/\s+/).reject(&:empty?).size
  chars = text.size
  puts "\e[36m📊 Статистика:\e[0m"
  puts "  Слов: #{words}"
  puts "  Символов: #{chars}"
  puts "  Время: #{elapsed.round(2)} сек"

  output_file = options[:output] || File.basename(image_path, ".*") + ".txt"
  File.write(output_file, text)
  puts "\e[32m💾 Сохранено в: #{output_file}\e[0m"
rescue => e
  puts "\e[31m❌ Ошибка: #{e.message}\e[0m"
  exit 1
end

File: ocr_reader.py


### 1. `ocr_reader.py` (Python)

```python
# ocr_reader.py — Python версия

import sys
import os
import time
import argparse
from PIL import Image
import pytesseract
from colorama import init, Fore, Style

init(autoreset=True)

def ocr_image(image_path, lang='eng'):
    """Распознаёт текст с изображения с помощью Tesseract."""
    try:
        image = Image.open(image_path)
        text = pytesseract.image_to_string(image, lang=lang)
        return text.strip()
    except Exception as e:
        return f"❌ Ошибка: {e}"

def save_result(text, output_path):
    """Сохраняет результат в файл."""
    with open(output_path, 'w', encoding='utf-8') as f:
        f.write(text)
    print(f"{Fore.GREEN}💾 Сохранено в: {output_path}")

def main():
    parser = argparse.ArgumentParser(description='OCR Reader')
    parser.add_argument('image', help='Путь к изображению')
    parser.add_argument('-l', '--lang', default='eng', help='Язык (по умолчанию eng)')
    parser.add_argument('-o', '--output', help='Файл для сохранения результата')
    args = parser.parse_args()

    if not os.path.exists(args.image):
        print(f"{Fore.RED}❌ Файл не найден: {args.image}")
        sys.exit(1)

    print(f"{Fore.CYAN}📖 OCR Reader (Python)")
    print(f"📂 Обработка: {args.image}")
    print("⏳ Распознавание...")

    start = time.time()
    text = ocr_image(args.image, args.lang)
    elapsed = time.time() - start

    if text.startswith("❌"):
        print(text)
        sys.exit(1)

    print(f"\n{Fore.GREEN}Результат:{Style.RESET_ALL}")
    print("─" * 50)
    print(text)
    print("─" * 50)

    words = len(text.split())
    chars = len(text)
    print(f"{Fore.CYAN}📊 Статистика:{Style.RESET_ALL}")
    print(f"  Слов: {words}")
    print(f"  Символов: {chars}")
    print(f"  Время: {elapsed:.2f} сек")

    if args.output:
        save_result(text, args.output)
    else:
        default_output = os.path.splitext(args.image)[0] + ".txt"
        save_result(text, default_output)

if __name__ == "__main__":
    main()
